engine/src/main/java/com/datatorrent/stram/util/SharedPubSubWebSocketClient.java

```diff
@@ -57,21 +57,33 @@
 
   }
 
+  /**
+   * Construct a SharedPubSubWebSocketClient with the given parameters
+   * @param uri The web socket server uri
+   * @param timeoutMillis The connection timeout
+   */
   public SharedPubSubWebSocketClient(URI uri, long timeoutMillis)
   {
     this.setUri(uri);
     lastConnectTryTime = System.currentTimeMillis();
     this.timeoutMillis = timeoutMillis;
   }
 
+  /**
+   * Construct a SharedPubSubWebSocketClient with the given parameters
+   * @param uri The web socket server uri as string
+   * @param timeoutMillis The connection timeout
+   */
   public SharedPubSubWebSocketClient(String uri, long timeoutMillis) throws URISyntaxException
   {
     this(new URI(uri), timeoutMillis);
   }
 
-  public void setup() throws IOException, ExecutionException, InterruptedException, TimeoutException
+  public synchronized void openConnection() throws IOException, ExecutionException, InterruptedException, TimeoutException
   {
-    openConnection(timeoutMillis);
+    if (!isConnectionSetup()) {
```

Comment:
Can you move the check to the super class? Is there a reason to avoid the same check in PubSubWebSocketClient?

Reply:
Since it is not shared.. I think it is reasonable that it expect that the caller adheres to a prescribed lifecycle.

Reply:
Why rely on a caller when it will not cost anything to do check in PubSubWebSocketClient (it does not need to enforce concurrency). To do check in SharedPubSubWebSocketClient a new method is required, that otherwise is not needed.

Reply:
In the case state consistency needs to happen everywhere in PubSubWebSocketClient not just in the openConnection method, for example publish should check if connection is available etc. Doing it just for one method is not correct, even in that case if you want other classes to extend the functionality it is very likely they would need to know the state of the client and while this method is not exactly a get the current state, it is in the vein and can be changed later to a more generic state method if we have the need to implement state consistency across the board.

Reply:
Does not 'publish()' in PubSubWebSocketClient already checks for connection in assertUsable()?

I would agree to introduce the new method when the need is identified. For now, it is not needed.

Reply:
I thought it was only looking for exceptions, looks like it is doing more than that. To be consistent, openConnection should also be throwing an exception and there would be this wonky code handling the exception in the SharedPubSub. If you look at SharedPubSubWebSocketClient, it is using the method isConnectionOpen in publish for example, so it may not be a bad to have it use isConnectionSetup as well.

Reply:
Sorry, I don't follow. I think that check for 'connection == null' should be part of 'PubSubWebSocketClient.openConnection()' to prevent double invocation. Once that check is moved there, there is no use case for 'isConnectionSetup()'. I don't see how moving check to PubSubWebSocketClient affects exception handling (I do not suggest to use assertUsable() in Shared).

Reply:
No, I didn't mean that. Other methods check for correct state otherwise they throw an exception like publish throws exception if openConnection hasn't been called. In the same vein, openConnection should throw an exception if called again because it is not in the correct state (pre-open) when it is called again. Just like a method is available to check the connection state, that shared, for example, is calling before calling publish, this method provides a similar functionality. Checking for connection null in openConnection in PubSub and a no-op would not be consistent with how the rest of the methods operate. I think the extra method will be more consistent with how things are implemented currently in that class.

Reply:
I don't think it is breaking oop principals that a second call to open connection cannot be made successfully if one is already setup. In fact, it is not a good thing for the second PubSub openConnection to just return trivially. It depends on where you draw the boundary of responsibility. It is not without precedent, there is a similar pattern with isConnectionOpen. These methods relay the current state of the class for the caller to know whether their call will be successful or if they need to make a remedy. For example a sub-class may want to close a previous open connection and open a new connection. I think this is a better approach and would like to leave it this way.

Reply:
I strongly disagree and prefer not to add isConnectionSetup().

Reply:
Let's agree to disagree on this one. I think you are LGTM as you mentioned, can you merge.